pyproject.toml

```diff
@@ -1,11 +1,86 @@
 [build-system]
 requires = [
-    "setuptools >= 35.0.2",
-    "wheel >= 0.29.0",
-    "incremental >= 21.3.0",
+  "setuptools >= 35.0.2",
+  "wheel >= 0.29.0",
+
+  # What follows is an exact copy of the runtime dependencies, because (A)
+  # incremental needs to be able to import `treq` to be able to import
+  # `treq.__version__` (see `tools.setuptools.dynamic` below), and (B) the
+  # `treq` module is set up such that everything is eagerly imported into the
+  # top-level namespace and thus sucks in every other dependency that we
+  # require.
+  "incremental >= 21.3.0",
+  "requests >= 2.1.0",
+  "hyperlink >= 21.0.0",
+  "Twisted[tls] >= 22.10.0",  # For #11635
+  "attrs",
+  "typing_extensions >= 3.10.0",
+  "multipart",
 ]
 build-backend = "setuptools.build_meta"
 
+[project]
+name = "Treq"
+dynamic = [
+    "version",
+]
+
+description = "High-level Twisted HTTP Client API"
+readme = "README.rst"
+license = { text = "MIT License" }
+# When updating this value, make sure our CI matrix includes a matching minimum version.
+requires-python = ">=3.8.0"
+authors = [
+  { name = "David Reid", email = "dreid@dreid.org" },
+]
+maintainers = [
+  { name = "Tom Most", email = "twm@freecog.net" },
+]
+classifiers = [
+    "Development Status :: 5 - Production/Stable",
+    "Intended Audience :: Developers",
+    "License :: OSI Approved :: MIT License",
+    "Operating System :: OS Independent",
+    "Framework :: Twisted",
+    "Programming Language :: Python",
+    "Programming Language :: Python :: 3.8",
+    "Programming Language :: Python :: 3.9",
+    "Programming Language :: Python :: 3.10",
+    "Programming Language :: Python :: 3.11",
+    "Programming Language :: Python :: 3.12",
+    "Programming Language :: Python :: Implementation :: CPython",
+    "Programming Language :: Python :: Implementation :: PyPy",
+]
+dependencies = [
+  "incremental >= 21.3.0",
+  "requests >= 2.1.0",
+  "hyperlink >= 21.0.0",
+  "Twisted[tls] >= 22.10.0",  # For #11635
+  "attrs",
+  "typing_extensions >= 3.10.0",
+  "multipart",
+]
+
+[project.optional-dependencies]
+dev = [
+    "pep8",
+    "pyflakes",
+    "httpbin==0.7.0",
+    "werkzeug==2.0.3",
+]
+
+docs = [
+    "sphinx<7.0.0",  # Removal of 'style' key breaks RTD.
+]
+
+[project.urls]
+Source = "https://github.com/twisted/treq"
+
+[tool.setuptools.packages.find]
+where = ["src"]
+[tool.setuptools.dynamic]
+version = {attr = "treq.__version__"}
```

Comment:
Oh, I see, you've tried to use setuptools `attr` and it's falling back to importing the module. 

I don't see anything here equivalent to the `use_incremental=True` argument to `setup()`. It looks like we need to add support to `incremental` for direct integration with setuptools, like setuptools_scm has:

https://github.com/pypa/setuptools_scm/blob/102f56a896bc4ea8d4c58d9c49f455f15310de38/pyproject.toml#L76-L77

Reply:
Yup.

Reply:
Filed https://github.com/twisted/incremental/issues/90 for this so that we can fix it in incremental. Sad that this seems to be the main setuptools build extension entrypoint.